pip/pip-289.md

```diff
@@ -0,0 +1,256 @@
+
+In order to add support for the existing `sensitive` annotation, I propose fixing all the connectors that have explicit secrets in their configurations.
+
+Fixing the documentation will be a matter of updating the existing documentation to reflect the current state of the code.
+
+## Public-facing Changes
+
+### Public API
+
+#### Add new method to SecretsProvider Interface
+
+Add the following method to the `SecretsProvider` interface:
+
+```java
+interface SecretsProvider {
+    /**
+     * If the passed value is formatted as a reference to a secret, as defined by the implementation, return the
+     * referenced secret. If the value is not formatted as a secret reference or the referenced secret does not exist,
+     * return null.
+     *
+     * @param value a config value that may be formatted as a reference to a secret
+     * @return the materialized secret. Otherwise, null.
+     */
+    default String interpolateSecretForValue(String value) {
+        return null;
+    }
+}
+```
+
+There are only two official implementations of the `SecretProvider` interface. The `ClearTextSecretsProvider` and the `EnvironmentBasedSecretsProvider`. Given that the `ClearTextSecretsProvider` is only plaintext, it will not override the new method. Here is the proposed implementation for the `EnvironmentBasedSecretsProvider`:
+
+```java
+public class EnvironmentBasedSecretsProvider implements SecretsProvider {
+    /**
+     * Pattern to match ${secretName} in the value.
+     */
+    private static final Pattern interpolationPattern = Pattern.compile("\\$\\{(.+?)}");
+
+    @Override
+    public String interpolateSecretForValue(String value) {
+        Matcher m = interpolationPattern.matcher(value);
+        if (m.matches()) {
+            String secretName = m.group(1);
+            // If the secret doesn't exist, we return null and don't override the current value.
+            return provideSecret(secretName, null);
+        }
+        return null;
+    }
+}
+```
+
+### Binary protocol
+
+No change.
+
+### Configuration
+
+Add `interpolateSecretsIntoConfigMap` to the `WorkerConfig`. This configuration will allow operators to determine whether users can interpolate secrets into configurations. Here is the potential documentation for this flag:
+
+```java
+class WorkerConfig {
+    @FieldContext(
+            category = CATEGORY_CONNECTORS,
+            doc = "Whether to interpolate secrets into the connector's configuration. Only affects Sinks and Sources."
+                    + " Applies to all connectors deployed by this function worker."
+                    + " When true, the SecretsProvider will recursively interpolate secrets into the connector's "
```

Comment:
What would we do with this configuration? Can you show you an overview of the implementation code?

> When true, the SecretsProvider will recursively interpolate secrets into the connector's .

Is the `SecretsProvider` implemented by the user himself?

Reply:
Please see https://github.com/apache/pulsar/pull/20901. The `SecretsProvider` is somewhat configurable (though the code is not very straightforward).